pos_payment_method_cashdro/static/src/js/screens.js

```diff
@@ -0,0 +1,243 @@
+                    console.log(res);
+                    operation_id = res.data;
+                    _this.pos.get_order().cashdro_operation = operation_id;
+                    // Acknowledge the operation
+                    var ack_url = _this._cashdro_ack_url(operation_id);
+                    return _this._cashdro_request(ack_url);
+                })
+                .then(function (res_ack) {
+                    // Validate the operation
+                    console.log(res_ack);
+                    var ask_url = _this._cashdro_ask_url(operation_id);
+                    return _this._cashdro_request_payment(ask_url);
+                })
+                .then(function (operation_data) {
+                    // This might be too verbose, but it helps a lot to diagnose
+                    // issues and their reasons.
+                    console.log(operation_data);
+                    var data = JSON.parse(operation_data.data);
+                    payment_line.cashdro_operation_data = data;
+                    _this.pos.get_order().in_cashdro_transaction = false;
+                    var tendered = data.operation.totalin / 100;
+                    payment_line.set_amount(tendered)
+                });
+            return request;
+        },
+
+        cashdro_finish_operation: function (operation) {
+            // Finish the Cashdro running operation
+            var _this = this
+            var order = this.pos.get_order()
+            if (operation) {
+                this._cashdro_request(this._cashdro_finish_url(operation))
+                .then(function () {
+                    order.in_cashdro_transaction = false;
+                    order.cashdro_operation = false;
+                    order.in_cashdro_transaction = false;
+                    _this.order_changes();
+                });
+            }
+        },
+
+        // API communication methods
+
+        _cashdro_url: function () {
+            // Cashdro machines don't support safe POST calls, so we're sending
+            // all the data quite unsafely constantly...
+            var cashdro_host = this.journal && this.journal.cashdro_host;
+            if (!cashdro_host) {
+                return false;
+            }
+            var url = "https://" + cashdro_host + "/Cashdro3WS/index.php?";
+            url += "name=" + this.journal.cashdro_user;
+            url += "&password=" + this.journal.cashdro_password;
+            return url;
+        },
+
+        _cashdro_payment_url: function(parameters) {
+            // Compose the url for a sale report to Cashdro
+            var url = this._cashdro_url()
+            url += "&operation=startOperation&type=4";
+            url += "&posid=pos-" + this.pos.pos_session.name;
+            url += "&posuser=" + this.pos.get_cashier().id;
+            url += "&parameters=" + JSON.stringify(parameters);
+            return url;
+        },
```

Comment:
Just a suggestion: [$.param](https://api.jquery.com/jquery.param/) for readability
Could do something like this, for example

```suggestion
        /**
         * Compose the url for a sale report to Cashdro
         * @param {Object} parameters
         * @returns {String}
         */
        _cashdro_payment_url: function(parameters) {
            return [
                this._cashdro_url(),
                $.param({
                    operation: "startOperation",
                    type: 4,
                    posid: "pos-" + this.pos.pos_session.name,
                    posuser: this.pos.get_cashier().id,
                    parameters: JSON.stringify(parameters),
                }),
            ].join("&")
        },
```

Reply:
Nice!